### DISCUSSION TITLE
Add Support for `.config` as a Config Search Location

### DISCUSSION THREAD

**Discussion by sairus2k** [eslint/eslint#17623] [Ideas] [answered]
I'd like to propose an enhancement to eslint configuration search mechanism. Currently, ESLint searches for the configuration file `./eslint.config.js` in the root directory of the project. I propose that we also include `./.config/eslint.js` as a valid location for the ESLint configuration file.

This proposal aligns with the [dot-config initiative](https://dot-config.github.io/), which advocates for a standardized approach to application configuration by using a `.config` directory in the root of the project. This approach provides a more organized structure for project configuration files, making it easier for developers to manage and navigate their projects.

Notably, this proposal differs from a previously discussed issue (https://github.com/eslint/eslint/issues/6008), which suggested adding `$HOME/.config` as a configuration search location. This proposal is focused on adding support for the project root `.config` directory, not the home directory.

It's worth mentioning that tools using [`cosmiconfig`](https://github.com/cosmiconfig/cosmiconfig) or [`lilconfig`](https://github.com/antonk52/lilconfig) for configuration search already support the `.config` directory. A notable example is `Stylelint`.

**Comment by Rec0iL99** [comment, marked as answer]
Hi @sairus2k, when linting via the CLI you can use the [`--config`](https://eslint.org/docs/latest/use/command-line-interface#-c---config) option to specify the config file in a custom location. ESLint will not perform a further search for config files.

```bash
npx eslint -c .config/eslint.config.js file.js
```

**Reply by sairus2k** [reply]
Thank you for the response. While the `--config` option does provide a workaround, it also introduces a significant amount of overhead. As you pointed out, it requires updating CI/CD configurations, commit-hook tools like lint-staged or lefthook, IDE settings, and also necessitates additional instructions for contributors. This added complexity can potentially act as a barrier to contribution and increase the maintenance burden for the project.

The main motivation behind this proposal is to reduce clutter in the project root directory. Currently, the root directory of many JavaScript projects is often filled with various configuration files, which can make it harder to navigate and understand the project structure. By supporting `./.config/eslint.js` as a valid location for the ESLint configuration file, we can help make project directories cleaner and more organized.

This approach is not about changing the way we have always done things, but rather about improving the developer experience. It provides a more organized structure for project configuration files, making it easier for developers to manage and navigate their projects.

I believe this change aligns with the philosophy of ESLint, which is to provide a pluggable and configurable linter that developers can tailor to their specific needs. Adding support for the `.config` directory would be another step in this direction, offering developers more flexibility in how they structure their projects.

I look forward to hearing your thoughts on this.

**Reply by Rec0iL99** [reply]
> which is to provide a pluggable and configurable linter that developers can tailor to their specific needs

Yes, ESLint is not strict in making you **only** have the option of putting the config file in the project's root. You can put your config in a `.config/` directory and tell ESLint to look there and ESLint will do just that.

**Reply by Rec0iL99** [reply]
I believe https://github.com/eslint/eslint/discussions/17623#discussioncomment-7211052 was marked as an answer by Nicholas. It got unmarked somehow. Going to mark it again.
